lib/iris/tests/unit/lazy_data/test_convert_nans_array.py

```diff
@@ -0,0 +1,126 @@
+        result = convert_nans_array(array)
+        self.assertNotIsInstance(result, ma.MaskedArray)
+        self.assertIs(result, array)
+
+    def test_nans_masked(self):
+        result = convert_nans_array(self.array,
+                                    nans_replacement=ma.masked)
+        self.assertIsInstance(result, ma.MaskedArray)
+        self.assertArrayEqual(result.mask, [[False, True],
+                                            [False, False]])
+        dummy = 666.0
+        result[0, 1] = dummy
+        self.assertArrayEqual(result.data, [[1.0, dummy],
+                                            [3.0, 4.0]])
+        self.assertIsNot(result, self.array)
+        # Check that fill value is the "standard" one for the type.
+        expected = ma.masked_array(self.array[0], dtype=self.array.dtype)
+        self.assertEqual(result.fill_value, expected.fill_value)
+
+    def test_nans_filled(self):
+        fill_value = 666.0
+        result = convert_nans_array(self.array,
+                                    nans_replacement=fill_value)
+        self.assertNotIsInstance(result, ma.MaskedArray)
+        self.assertIs(result, self.array)
+        expected = np.array([[1.0, fill_value],
+                             [3.0, 4.0]])
+        self.assertArrayEqual(result, expected)
+
+    def test_nans_filled_failure(self):
+        fill_value = 1e+20
+        dtype = np.dtype('int16')
+        emsg = 'Fill value of .* invalid for array result .*'
+        with self.assertRaisesRegexp(ValueError, emsg):
+            convert_nans_array(self.array,
+                               nans_replacement=fill_value,
+                               result_dtype=dtype)
+
+    def test_nans_none_failure(self):
+        emsg = 'Array contains unexpected NaNs'
+        with self.assertRaisesRegexp(ValueError, emsg):
+            convert_nans_array(self.array)
+
+    def test_result_dtype_cast_float_to_int(self):
+        dtype = np.int
+        result = convert_nans_array(self.array,
+                                    nans_replacement=ma.masked,
+                                    result_dtype=dtype)
+        self.assertIsInstance(result, ma.MaskedArray)
+        expected = ma.masked_array([[1, 2],
+                                    [3, 4]],
+                                   dtype=dtype)
+        expected[0, 1] = ma.masked
+        self.assertArrayEqual(result, expected)
+        self.assertIsNot(result, self.array)
+
+    def test_filled_result_dtype_cast_float_to_int(self):
+        fill_value = 666
+        dtype = np.int
+        result = convert_nans_array(self.array,
+                                    nans_replacement=fill_value,
+                                    result_dtype=dtype)
+        self.assertNotIsInstance(result, ma.MaskedArray)
+        expected = np.array([[1, fill_value],
+                             [3, 4]],
+                            dtype=dtype)
+        self.assertArrayEqual(result, expected)
```

Comment:
In this case, result is **not** the original (not in-place with new dtype).

P.S. did you know you can *assign* `array.dtype` ?  It does like a C pointer cast :scream: .